### PR TITLE
[tir] make IterVar extends Var

### PR DESCRIPTION
**Pull request by yzhliu**
@tqchen @ZihengJiang @Hzfengsy Please review

### CODE REVIEW

File: src/tir/ir/expr.cc [hunk 1]
```diff
@@ -45,24 +45,28 @@ SizeVar::SizeVar(std::string name_hint, DataType t)
...
-                          Var var,
-                          IterVarType t,
-                          std::string thread_tag) {
-  ObjectPtr<IterVarNode> n = make_object<IterVarNode>();
-  n->dom = dom;
-  n->var = var;
-  n->iter_type = t;
-  n->thread_tag = thread_tag;
-  return IterVar(n);
+IterVarNode::IterVarNode(DataType dtype, std::string name_hint,
```
Comment (Hzfengsy) [inline src/tir/ir/expr.cc]: I'm not sure whether we need a constructor for Node

File: include/tvm/tir/expr.h [hunk 2]
```diff
@@ -236,21 +228,33 @@ enum IterVarType : int {
...
+   */
+  TVM_DLL IterVar(Range dom, IterVarType iter_type,
+                  std::string name_hint = "v",
+                  DataType t = DataType::Int(32),
+                  std::string thread_tag = "");
   /*!
    * \brief access the internal node container
    * \return the pointer to the internal node container
    */
   inline const IterVarNode* operator->() const;
```
Comment (tqchen) [inline include/tvm/tir/expr.h]: We can likely use TVM_DEFINE_OBJECT_REF_METHODS to define these methods

File: include/tvm/tir/expr.h [hunk 3]
```diff
@@ -236,21 +228,33 @@ enum IterVarType : int {
...
-  explicit IterVar(ObjectPtr<Object> n) : ObjectRef(n) {}
+  explicit IterVar(ObjectPtr<Object> n) : Var(n) {}
+  /*! \brief constructor.
+   * \param dom interval of the variable.
+   * \param iter_type indicate the iteration type of the variable.
+   * \param name_hint variable name
+   * \param t data type
+   * \param thread_tag additional tag on the iteration variable.
+   */
+  TVM_DLL IterVar(Range dom, IterVarType iter_type,
```
Comment (tqchen) [inline include/tvm/tir/expr.h]: let us cross reference python API to check if it is the right order

Comment (yzhliu) [reply include/tvm/tir/expr.h]: in python it's `_IterVar(dom, name, iter_type, thread_tag='')`, while if we need to make `name` optional, we have to make `name` after `iter_type`, right?

File: src/tir/ir/expr.cc (hunk 1 above)
Comment (yzhliu) [reply src/tir/ir/expr.cc]: it's used in https://github.com/apache/incubator-tvm/pull/4765/files#diff-b0d470d0765d89448d8d18506ba0c54eR61

File: include/tvm/tir/expr.h (hunk 3 above)
Comment (tqchen) [reply include/tvm/tir/expr.h]: I feel in this case perhaps it is also worth to consider consistency withVar, which means we should put name first. But provide auxilary factory functions(e.g. thread_axis) to quickly construct some of the data

**Comment by yzhliu** [comment]
@tqchen another problem is currently we rely on `IterVar->var` to propagate variables from the very beginning, e.g., 
```
rv = tvm.reduce_axis(...)
Y = sum(X[rv], axis=rv)
```

However, `rv` could be replaced during the pass, it cause problem when we use the original `rv` to access array 

one example of replacing: https://github.com/apache/incubator-tvm/blob/master/src/tir/ir/expr_functor.cc#L206-L208

**Comment by tqchen** [comment]
It is indeed tradeoff in here, it is similar to early binding of information vs late binding. @yzhliu If you think it is a good idea to keep IterVar as it is(to simplify the mutation of vars), we can close this PR for now.

**Comment by yzhliu** [comment]
Thanks. I'll close it for now.

**Pull request closed without merging**